packetbeat/flows/flowid.go

```diff
@@ -0,0 +1,482 @@
+package flows
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"net"
+)
+
+type FlowID struct {
+	rawFlowID
+	flow Flow // remember associated flow for faster lookup
+}
+
+type rawFlowID struct {
+	flowID []byte
+	flowIDMeta
+	dir flowDirection
+}
+
+type flowIDMeta struct {
+	flags FlowIDFlag
+
+	// offsets into flowID
+	offEth        uint8
+	offOutterVlan uint8
+	offVlan       uint8
+	offOutterIPv4 uint8
+	offIPv4       uint8
+	offOutterIPv6 uint8
+	offIPv6       uint8
+	offICMPv4     uint8
+	offICMPv6     uint8
+	offUDP        uint8
+	offTCP        uint8
+	offID         uint8
+
+	cntEth  uint8
+	cntVlan uint8
+	cntIP   uint8
+}
+
+type FlowIDFlag uint16
+
+const (
+	EthFlow        FlowIDFlag = (1 << 0)
+	OutterVlanFlow FlowIDFlag = (1 << 1)
+	VLanFlow       FlowIDFlag = (1 << 2)
+	OutterIPv4Flow FlowIDFlag = (1 << 3)
+	IPv4Flow       FlowIDFlag = (1 << 4)
+	OutterIPv6Flow FlowIDFlag = (1 << 5)
+	IPv6Flow       FlowIDFlag = (1 << 6)
+	ICMPv4Flow     FlowIDFlag = (1 << 7)
+	ICMPv6Flow     FlowIDFlag = (1 << 8)
+	UDPFlow        FlowIDFlag = (1 << 9)
+	TCPFlow        FlowIDFlag = (1 << 10)
+	ConnectionID   FlowIDFlag = (1 << 11)
```

Comment:
Why not use iota for these? See Effective Go for an example of using iota for powers of two.